lib/mocha/parameter_matchers/positional_or_keyword_hash.rb

```diff
@@ -9,14 +10,20 @@ def initialize(value)
         @value = value
       end
 
-      def matches?(available_parameters)
+      def matches?(available_parameters) # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
         parameter, is_last_parameter = extract_parameter(available_parameters)
         return false unless parameter.is_a?(Hash)
 
-        if is_last_parameter && Mocha.configuration.strict_keyword_argument_matching?
-          return false unless ::Hash.ruby2_keywords_hash?(parameter) == ::Hash.ruby2_keywords_hash?(@value)
+        equal = parameter == @value
+        if is_last_parameter && Mocha::RUBY_V27_PLUS
+          same_type_of_hash = same_type_of_hash?(parameter, @value)
+          if Mocha.configuration.strict_keyword_argument_matching?
+            return false unless same_type_of_hash
+          elsif equal && !same_type_of_hash
+            Deprecation.warning("#{mocha_inspect} will stop matching #{parameter.mocha_inspect} when strict keyword argument matching is enabled.")
```

Comment:
Yeah. I wonder if that's a wider problem too, i.e. with expectation matching...? But perhaps in that case it's clearer because the hashes (and other arguments) are all within parentheses...?

Reply:
Let me try to walk through what happened in my mind, in case it helps:
- in the test output, I got `Mocha deprecation warning at /path/to/positional_or_keyword_hash_test.rb:38:in 'test_should_match_hash_arg_with_keyword_args_but_display_deprecation_warning_if_appropriate': :key_1 => 1, :key_2 => 2 will stop matching {:key_1 => 1, :key_2 => 2} when strict keyword argument matching is enabled.`
- line 38 was `assert matcher.matches?([{ :key_1 => 1, :key_2 => 2 }])`
  - I don't think I would have understood how the deprecation warning related to this if I didn't already have the context of https://github.com/freerange/mocha/pull/562.
  - For a typical user, I guess they would see an actual method invocation e.g. `foo({ key_1: 1})`. Thus, maybe just highlighting that this method expected a kwarg, and got a positional hash would be sufficient.

Reply:
Thanks. I've attempted to address this in https://github.com/freerange/mocha/pull/563/commits/061453c8e5548e1fe08d4f70fdd152e784f09db0.

Reply:
Example deprecation warnings:

> Expected keyword arguments (:key_1 => 1, :key_2 => 2), but received positional hash ({:key_1 => 1, :key_2 => 2}). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.

OR

> Expected positional hash ({:key_1 => 1, :key_2 => 2}), but received keyword arguments (:key_1 => 1, :key_2 => 2). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.

Reply:
So long, but every word seems important 😄 I think this is much clearer, hopefully other devs will feel the same way too.

Reply:
Yeah, I think I prefer to err on the side of clarity! Many thanks for your help on improving this!

Reply:
And you can always disable deprecation warnings if you hate them that much!